include/visitors/code_gen.h

```diff
@@ -2,30 +2,545 @@
 
 #include "binaryen-c.h"
 
-void test()
+class CodeGen : public Visitor
 {
-    BinaryenModuleRef mod = BinaryenModuleCreate();
-
-    BinaryenType ii[2] = {BinaryenTypeInt32(), BinaryenTypeInt32()};
-    BinaryenType params = BinaryenTypeCreate(ii, 2);
-    BinaryenType results = BinaryenTypeInt32();
-
-    BinaryenExpressionRef x = BinaryenLocalGet(mod, 0, BinaryenTypeInt32());
-    BinaryenExpressionRef y = BinaryenLocalGet(mod, 1, BinaryenTypeInt32());
-    BinaryenExpressionRef add = BinaryenBinary(mod, BinaryenAddInt32(), x, y);
-
-    BinaryenFunctionRef adder = BinaryenAddFunction(mod, "adder", params, results, NULL, 0, add);
-    BinaryenAddFunctionExport(mod, "adder", "adder");
-
-    BinaryenModulePrint(mod);
-}
-
-/*
- * NOTE:
- *   This has been commented out because LLVM was removed as a dependency from the project.
- *   We will be using Binaryen instead, but this code is still useful for reference.
- *   Once the Binaryen implementation is complete, this code will be removed.
- */
+    std::shared_ptr<Environment<BinaryenExpressionRef>> environment;
+    std::map<std::string, std::string> std_lib;
+    Stack<BinaryenExpressionRef> stack;
+
+    std::vector<BinaryenExpressionRef> fn_body;
+    std::vector<BinaryenType> locals;
+
+    BinaryenModuleRef mod;
+    int local_index;
+
+public:
+    ~CodeGen()
+    {
+        BinaryenModuleDispose(this->mod);
+    }
+
+    CodeGen() : mod(BinaryenModuleCreate()), local_index(0)
+    {
+        this->environment =
+            std::make_shared<Environment<BinaryenExpressionRef>>();
+    }
+
+    void init_std_lib()
+    {
+        // format string and pointer as param for printf
```

Comment:
The environment will inject JavaScript's `console.log` function instead of c's `printf` function, so the parameters and results will be different.